ledger/src/check_next_block.rs

```diff
@@ -129,6 +132,40 @@ impl<N: Network, C: ConsensusStorage<N>> Ledger<N, C> {
         Ok(())
     }
 
+    /// Check that leaves in the subdag point to batches in other blocks that are valid.
+    fn check_block_subdag_leaves(&self, block: &Block<N>) -> Result<()> {
+        // Check if the block has a subdag.
+        let Authority::Quorum(subdag) = block.authority() else {
+            return Ok(());
+        };
+
+        // Store the IDs of all certificates in this subDAG.
+        // This allows determining which edges point to other subDAGs/blocks.
+        let subdag_certs: HashSet<_> = subdag.certificate_ids().collect();
+        let leaf_edges: HashSet<_> = subdag
+            .certificates()
+            .flat_map(|cert| cert.previous_certificate_ids().iter().map(|id| (cert.round() - 1, id)))
+            .filter(|(_, id)| !subdag_certs.contains(id))
+            .collect();
+
+        for (round, cert_id) in leaf_edges {
+            // Find the block for this certificate.
+            // We might check the same block multiple times, but, so far, the check
+            // simply ensures that the block is a valid ancestor.
+            match self.vm.block_store().get_block_for_certificate(cert_id, Some(round))? {
```

Comment:
I don't think we need to test the round, because elsewhere we check that each certificate in the sub-DAG has a quorum of signers, which means (under the usual fault tolerance assumptions) that at least one correct validator has ensured that the certificate's previous certificate IDs identify certificates in the round just before the certificate's round.

Reply:
Yes, that makes sense. I will remove the check and add a note here.